crates/holochain/src/core/workflow/sys_validation_workflow/unit_tests.rs

```diff
@@ -0,0 +1,285 @@
+    let response = WireOps::Record(ops);
+    network
+        .expect_get()
+        .return_once(move |_, _| Ok(vec![response]));
+
+    test_case.with_network_behaviour(network).run().await;
+
+    let ops_to_app_validate = test_case.get_ops_pending_app_validation().await;
+    assert!(ops_to_app_validate.contains(&op_hash));
+
+    test_case.expect_app_validation_triggered().await;
+}
+
+struct TestCase {
+    dna_def: DnaDef,
+    dna_hash: DnaDefHashed,
+    test_space: TestSpace,
+    keystore: MetaLairClient,
+    agent: AgentPubKey,
+    app_validation_trigger: (TriggerSender, TriggerReceiver),
+    self_trigger: (TriggerSender, TriggerReceiver),
+    actual_network: Option<MockHolochainP2pDnaT>,
+}
+
+impl TestCase {
+    async fn new() -> Self {
+        let dna_def = DnaDef::unique_from_zomes(vec![], vec![]);
+        let dna_hash = DnaDefHashed::from_content_sync(dna_def.clone());
+
+        let test_space = TestSpace::new(dna_hash.hash.clone());
+
+        let keystore = holochain_keystore::test_keystore();
+        let agent = keystore.new_sign_keypair_random().await.unwrap().into();
+
+        Self {
+            dna_def,
+            dna_hash,
+            test_space,
+            keystore,
+            agent,
+            app_validation_trigger: TriggerSender::new(),
+            self_trigger: TriggerSender::new(),
+            actual_network: None,
+        }
+    }
+
+    fn dna_hash(&self) -> DnaHash {
+        self.dna_hash.hash.clone()
+    }
+
+    fn dht_db_handle(&self) -> DbWrite<DbKindDht> {
+        self.test_space.space.dht_db.clone()
+    }
+
+    fn cache_db_handle(&self) -> DbWrite<DbKindCache> {
+        self.test_space.space.cache_db.clone()
+    }
+
+    async fn sign_action(&self, action: Action) -> SignedActionHashed {
+        let action_hashed = ActionHashed::from_content_sync(action);
+        SignedActionHashed::sign(&self.keystore, action_hashed)
+            .await
+            .unwrap()
+    }
+
+    fn with_network_behaviour(&mut self, network: MockHolochainP2pDnaT) -> &mut Self {
+        self.actual_network = Some(network);
+        self
+    }
+
+    async fn save_op_to_db<T: DbKindT>(
+        &self,
+        db: DbWrite<T>,
+        op: DhtOp,
+    ) -> StateMutationResult<DhtOpHash> {
+        let op = DhtOpHashed::from_content_sync(op);
+
+        let test_op_hash = op.as_hash().clone();
+        db.write_async({
+            move |txn| -> StateMutationResult<()> {
+                holochain_state::mutations::insert_op(txn, &op)?;
+                Ok(())
+            }
+        })
+        .await
+        .unwrap();
+
+        Ok(test_op_hash)
+    }
+
+    async fn run(&mut self) {
+        // TODO So this struct is just here to follow the 'workspace' pattern? The Space gets passed to the workflow anyway and most of the fields are shared.
+        //      Maybe just moving the Space to the workspace is enough to tidy this up?
+        let workspace = SysValidationWorkspace::new(
```

Comment:
So many things to clean up...the whole `Space` thing itself is an ill defined concept, and yet it's so pervasive. Maybe if it were renamed it would make more sense? Thoughts?

But yes this does seem unnecessary other than to follow the convention. And yes the whole "Space" could be passed in.

Reply:
I haven't understood what a space is. There are all the DB references in there. Mmmh. What is this ill defined pervasive concept?

Reply:
I've actually started refactoring this in the next PR. The space isn't heavily used and can be pulled up a few levels so it never needs to appear here, so this TODO will be addressed on an upcoming PR